pkg/os/disk/disk.go

```diff
@@ -47,51 +47,37 @@ const (
 func ListDiskLocations() (map[uint32]Location, error) {
 	// sample response
 	// [{
-	//    "number":  0,
-	//    "location":  "PCI Slot 3 : Adapter 0 : Port 0 : Target 1 : LUN 0"
+	//    "Index":  3,
+	//    "SCSILogicalUnit":  1,
+	//    "SCSITargetId":  0,
+	//    "SCSIPort":  1,
+	//    "SCSIBus":  0
 	// }, ...]
-	cmd := fmt.Sprintf("ConvertTo-Json @(Get-Disk | select Number, Location)")
+	cmd := fmt.Sprintf("ConvertTo-Json @(Get-CimInstance win32_diskdrive|where-object -FilterScript {$_.SCSIPort -Ne 0}|Select Index,SCSILogicalUnit,SCSITargetId,SCSIPort,SCSIBus)")
```

Comment:
`$_.SCSIPort -Ne 0` is not always correct for data disk because for v6 vm sku disk, SCSIPort of data is 0, 

Windows team provided the following info:
 - normal scsi data disk
`"Model":  "Microsoft Virtual Disk"`
 - v6 nvme controller data disk
 `"Model":  "Virtual_Disk NVME Premium"`

thus I think we could use this filter: `Where-Object { $_.Model -like "*Virtual?Disk*"`, I have worked out another PR to fix this issue permanently, and thanks for your contribution, I have forked your PR as first commit: https://github.com/kubernetes-sigs/azuredisk-csi-driver/pull/2671

Reply:
hm but that doesn't solve our issue - as outlined in the powershell cmdlet outputs we shared in the issue, Port is not always 0, and we use v5 machines.

the NVMe issue is also an issue and possibly related, but it's not the same issue.

Reply:
I think following command could solve all issues as I have tried on multiple VM skus including v5, v6:
```
ConvertTo-Json @(Get-CimInstance win32_diskdrive | Where-Object { $_.Model -like "*Virtual?Disk*" } | Select-Object Index, SCSILogicalUnit, SCSITargetId, SCSIPort, SCSIBus, Model)
```

judge by the `Model` field is more reliable, can you try on your v5 machine? thanks.

Reply:
v5 machine with ephemeral volume attached:

my pr (With added Model in select object):
```
> ConvertTo-Json @(Get-CimInstance win32_diskdrive|where-object -FilterScript {$_.SCSIPort -Ne 0}|Select Index,SCSILogicalUnit,SCSITargetId,SCSIPort,SCSIBus,Model)
[
    {
        "Index":  2,
        "SCSILogicalUnit":  0,
        "SCSITargetId":  0,
        "SCSIPort":  1,
        "SCSIBus":  0,
        "Model":  "Microsoft Virtual Disk"
    }
]

```

your suggestion:
```
> ConvertTo-Json @(Get-CimInstance win32_diskdrive | Where-Object { $_.Model -like "*Virtual?Disk*" } | Select-Object Index, SCSILogicalUnit, SCSITargetId, SCSIPort, SCSIBus, Model)
[
    {
        "Index":  1,
        "SCSILogicalUnit":  1,
        "SCSITargetId":  0,
        "SCSIPort":  0,
        "SCSIBus":  0,
        "Model":  "Microsoft Virtual Disk"
    },
    {
        "Index":  0,
        "SCSILogicalUnit":  0,
        "SCSITargetId":  0,
        "SCSIPort":  0,
        "SCSIBus":  0,
        "Model":  "Microsoft Virtual Disk"
    },
    {
        "Index":  2,
        "SCSILogicalUnit":  0,
        "SCSITargetId":  0,
        "SCSIPort":  1,
        "SCSIBus":  0,
        "Model":  "Microsoft Virtual Disk"
    }
]
```

Looking at `SCSILogicalUnit` we'll still have the issue of two disks with the same LUN ID and one is wrong (the one with Port 0) unfortunately. 

Not sure what other properties we could use which are more reliable though.

Reply:
I'm not sure if it makes a difference but we use Gen 2 VMs.

Reply:
then what about this query `ConvertTo-Json @(Get-CimInstance win32_diskdrive | Where-Object { $_.Model -eq "Virtual_Disk NVME Premium" -or $_.SCSIPort -eq 0 } | Select-Object Index, SCSILogicalUnit, SCSITargetId, SCSIPort, SCSIBus, Model)`

Reply:
there is a little typo in my query, try with `ConvertTo-Json @(Get-CimInstance win32_diskdrive | Where-Object { $_.Model -eq "Virtual_Disk NVME Premium" -or $_.SCSIPort -Ne 0 } | Select-Object Index, SCSILogicalUnit, SCSITargetId, SCSIPort, SCSIBus, Model)`

Reply:
yes, that works :+1: